File: study/lua/src/app/scene/view/avatar/AvatarDetailSkillCell.lua
--
-- Author: Liangxu
-- Date: 2018-1-5 14:07:33

local ListViewCellBase = require("app.ui.ListViewCellBase")
local AvatarDetailSkillCell = class("AvatarDetailSkillCell", ListViewCellBase)
local HeroSkillActiveConfig = require("app.config.hero_skill_active")

function AvatarDetailSkillCell:ctor(data)
	self._data = data
	self._label = nil --技能描述文本

	local resource = {
		file = Path.getCSB("HeroDetailSkillCell", "hero"),
		binding = {

		},
	}
	AvatarDetailSkillCell.super.ctor(self, resource)
end

function AvatarDetailSkillCell:onCreate()
	local contentSize = self._panelBg:getContentSize()
	local height = contentSize.height

	local skillId = self._data.id
	local unlock = self._data.unlock
	local isActiveJoint = self._data.isActiveJoint
	local isShowJointTip = unlock and not isActiveJoint --是否显示合击提示
	local config = HeroSkillActiveConfig.get(skillId)
	if config then
		local skillIconRes = config.skill_icon
		local unlockDes = isShowJointTip and Lang.get("avatar_detail_skill_unlock", {level = unlock}) or ""
		local skillDes = "["..config.name.."] "..config.description
		self._imageSkillIcon:loadTexture(Path.getCommonIcon("skill", skillIconRes))
		if self._label == nil then
			self._label = ccui.RichText:create()
			self._label:setAnchorPoint(cc.p(0, 1))
			self._label:ignoreContentAdaptWithSize(false)
			self._label:setContentSize(cc.size(262, 0))
			self._panelBg:addChild(self._label)
		end
		local content = Lang.get("avatar_detail_skill_des", {des = skillDes, unlock = unlockDes})
		self._label:setRichTextWithJson(content)
		self._label:formatText()
		local desHeight = self._label:getContentSize().height + 15
		height = math.max(contentSize.height, desHeight)--上下各扩展5像素
		self._label:setPosition(cc.p(120, height - 5))
	else
		logError(string.format("hero_skill_active config can not find id = %d", self._skillId))
	end
	
	local size = cc.size(contentSize.width, height) 
	self:setContentSize(size)
	self._imageSkillBg:setPosition(cc.p(17, height - 1))
	self._imageBg:setContentSize(cc.size(contentSize.width, height - 2))
end

return AvatarDetailSkillCell

File: study/ts/script/scene/view/avatar/AvatarDetailSkillCell.ts
import ListViewCellBase from "../../../ui/ListViewCellBase";
import { G_ConfigLoader } from "../../../init";
import { ConfigNameConst } from "../../../const/ConfigNameConst";
import { Lang } from "../../../lang/Lang";
import UIHelper from "../../../utils/UIHelper";
import { Path } from "../../../utils/Path";
import CommonUI from "../../../ui/component/CommonUI";
import { RichTextExtend } from "../../../extends/RichTextExtend";

const { ccclass, property } = cc._decorator;

@ccclass
export default class AvatarDetailSkillCell extends ListViewCellBase{
    @property({
        type: cc.Node,
        visible: true
    })
    _panelBg: cc.Node = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageBg: cc.Sprite = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageSkillBg: cc.Sprite = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageSkillIcon: cc.Sprite = null;

    private _label: cc.RichText;
    private _data:any;

    public static path:string = 'avatar/AvatarDetailSkillCell';

    ctor(data) {
        this._data = data;
    }
    onCreate() {
        var contentSize = this._panelBg.getContentSize();
        var height = contentSize.height;
        var skillId = this._data.id;
        var unlock = this._data.unlock;
        var isActiveJoint = this._data.isActiveJoint;
        var isShowJointTip = unlock && !isActiveJoint;
        var config = G_ConfigLoader.getConfig(ConfigNameConst.HERO_SKILL_ACTIVE).get(skillId);//HeroSkillActiveConfig.get(skillId);
        if (config) {
            var skillIconRes = config.skill_icon;
            var unlockDes = isShowJointTip && Lang.get('avatar_detail_skill_unlock', { level: unlock }) || '';
            var skillDes = '[' + (config.name + ('] ' + config.description));
            this._imageSkillIcon.node.addComponent(CommonUI).loadTexture(Path.getCommonIcon('skill', skillIconRes))
            if (this._label == null) {
                var node = new cc.Node();
                this._label = node.addComponent(cc.RichText);
                this._label.node.setAnchorPoint(cc.v2(0, 1));
                this._label.maxWidth = 262;
                this._panelBg.addChild(this._label.node);
            }
            var content = Lang.get('avatar_detail_skill_des', {
                des: skillDes,
                unlock: unlockDes
            });
            //this._label.setRichTextWithJson(content);
            // this._label.formatText();
            RichTextExtend.setRichTextWithJson(this._label,content)
            var desHeight = this._label.node.getContentSize().height + 15;
            height = Math.max(contentSize.height, desHeight);
            this._label.node.setPosition(cc.v2(120, height - 5));
        } else {
            //logError(string.format('hero_skill_active config can not find id = %d', this._skillId));
        }
        var size = cc.size(contentSize.width, height);
        this.node.setContentSize(size);
        this._imageSkillBg.node.setPosition(cc.v2(17, height - 1));
        this._imageBg.node.setContentSize(cc.size(contentSize.width, height - 2));
    }
}
